packages/addon/src/background.ts

```diff
@@ -421,9 +430,37 @@ async function addThundermailToken(
   }
 }
 
+function initStorageWatcher() {
+  browser.storage.onChanged.addListener(async (changes, area) => {
+    if (changes[STORAGE_KEY_AUTH]) {
+      if (changes[STORAGE_KEY_AUTH].newValue === undefined) {
+        try {
+          // OIDC logout
+          await authStore.logoutFromOIDC();
+        } catch (error) {
+          // We really shouldn't be using the authStore here, since we don't have the right
+          // env vars, but that does not interfere with logout.
+        }
+
+        try {
+          // FXA/JWT logout
+          await api.removeAuthToken();
+        } catch (error) {
+          console.error('Legacy (fxa) logout failed:', error);
+        }
+      }
+      // Optionally, we can check if they've just logged in:
+      // typeof changes[STORAGE_KEY_AUTH].newValue === 'object'
+      // In which case, we could shift some logic from the onMessage handler
+      // to this function.
+    }
+  });
+}
+
 (async function main() {
   initMenu();
   initCloudFile();
+  initStorageWatcher();
```

Comment:
Logging out of hamburger logs out of add-on, part 3/4

We call our new `initStorageWatcher()` as part of our initialization routine.